File: frontend/src/types/task.ts
export type TaskPriority = 'high' | 'medium' | 'low'
export type TaskCategory = 'judicial' | 'administrativa' | 'parecer' | 'publicacao_dje' | 'prazo'

export interface ChecklistItem {
  text: string
  done: boolean
}

export interface Task {
  id: string
  column_id: string
  title: string
  description?: string
  category?: TaskCategory
  priority: TaskPriority
  assigned_to?: string
  due_date?: string
  start_date?: string
  estimated_hours?: number
  tags?: string[]
  checklist?: ChecklistItem[]
  position: number
  source_ref?: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface Column {
  id: string
  board_id: string
  name: string
  position: number
  wip_limit?: number
  tasks: Task[]
}

export interface Board {
  id: string
  name: string
  columns: Column[]
}

export interface TaskCreatePayload {
  column_id: string
  title: string
  description?: string
  category?: TaskCategory
  priority?: TaskPriority
  assigned_to?: string
  due_date?: string
  start_date?: string
  estimated_hours?: number
  tags?: string[]
  checklist?: ChecklistItem[]
}

export interface TaskUpdatePayload {
  title?: string
  description?: string
  category?: TaskCategory
  priority?: TaskPriority
  assigned_to?: string | null
  due_date?: string | null
  start_date?: string | null
  estimated_hours?: number | null
  tags?: string[]
  checklist?: ChecklistItem[]
}

export interface TaskMovePayload {
  column_id: string
  position: number
}

export interface TaskHistoryEntry {
  id: string
  task_id: string
  from_column_id?: string
  to_column_id?: string
  changed_by?: string
  notes?: string
  created_at: string
}

export interface TaskComment {
  id: string
  task_id: string
  user_id: string
  content: string
  created_at: string
  updated_at: string
}

export interface UserMinimal {
  id: string
  name: string
  email: string
}

File: backend/app/schemas/task.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, model_validator

from app.models.task import TaskCategory, TaskPriority


# ---------------------------------------------------------------------------
# Checklist item schema
# ---------------------------------------------------------------------------

class ChecklistItem(BaseModel):
    text: str
    done: bool = False


# ---------------------------------------------------------------------------
# Task CRUD
# ---------------------------------------------------------------------------

class TaskCreate(BaseModel):
    column_id: uuid.UUID
    title: str
    description: Optional[str] = None
    category: Optional[TaskCategory] = None
    priority: TaskPriority = TaskPriority.medium
    assigned_to: Optional[uuid.UUID] = None
    due_date: Optional[datetime] = None
    start_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    tags: Optional[list[str]] = None
    checklist: Optional[list[ChecklistItem]] = None
    source_ref: Optional[dict] = None


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    category: Optional[TaskCategory] = None
    priority: Optional[TaskPriority] = None
    assigned_to: Optional[uuid.UUID] = None
    due_date: Optional[datetime] = None
    start_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    tags: Optional[list[str]] = None
    checklist: Optional[list[ChecklistItem]] = None


class TaskMoveRequest(BaseModel):
    column_id: uuid.UUID
    position: int = 0


class TaskOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    column_id: uuid.UUID
    title: str
    description: Optional[str] = None
    category: Optional[TaskCategory] = None
    priority: TaskPriority
    assigned_to: Optional[uuid.UUID] = None
    due_date: Optional[datetime] = None
    start_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    tags: Optional[list[str]] = None
    checklist: Optional[list[ChecklistItem]] = None
    position: int
    source_ref: Optional[dict] = None
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Task history
# ---------------------------------------------------------------------------

class TaskHistoryOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    task_id: uuid.UUID
    from_column_id: Optional[uuid.UUID] = None
    to_column_id: Optional[uuid.UUID] = None
    changed_by: Optional[uuid.UUID] = None
    notes: Optional[str] = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Task comments
# ---------------------------------------------------------------------------

class TaskCommentCreate(BaseModel):
    content: str


class TaskCommentOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    task_id: uuid.UUID
    user_id: uuid.UUID
    content: str
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Board / Column
# ---------------------------------------------------------------------------

class ColumnOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    board_id: uuid.UUID
    name: str
    position: int
    wip_limit: Optional[int] = None
    tasks: list[TaskOut] = []

    @model_validator(mode="after")
    def _sort_tasks(self) -> "ColumnOut":
        self.tasks.sort(key=lambda t: t.position)
        return self


class BoardOut(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    columns: list[ColumnOut] = []

    @model_validator(mode="after")
    def _sort_columns(self) -> "BoardOut":
        self.columns.sort(key=lambda c: c.position)
        return self


# ---------------------------------------------------------------------------
# User (lightweight, for listing assignees)
# ---------------------------------------------------------------------------

class UserMinimal(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    name: str
    email: str
